wordpress-mobile/WordPress-iOS
# Jetpack app: add it as a new project

This PR is part of an exploration to try different ways of building the Jetpack app.

This differs from https://github.com/wordpress-mobile/WordPress-iOS/pull/16095 because it adds a new project (`pbxproj`), instead of keeping the new app as just a new target/scheme.

This adds more complexity, configuration, and files. But the pros and cons are going to be discussed in the end.

<img src="https://user-images.githubusercontent.com/7040243/111365711-95f1a480-8671-11eb-8ea1-dc24f66b1a16.gif" width="300">

### To test

1. `pod install` (there are a few warnings due to `Generated duplicate UUIDs` that I haven't investigated why it's happening)
2. Build the WordPress scheme, it should work as usual
3. Build the Jetpack scheme, it should build and it should have a 1) different icon 2) different splash screen
4. Also, notice that we have two projects now: Jetpack and WordPress

### Testing different features

1. In the WordPress app, open the site list: nothing should happen
2. In the Jetpack app, open the site list: you should see a message saying "Welcome to the Jetpack app!" at the bottom

The different behavior was done using a preprocessor macro:

```swift
    #if JETPACK
    [self displayNoticeWithTitle:@"Welcome to the Jetpack app!" message:nil];
    #endif
```

### Pros

* Having Jetpack as a separate project helps to keep `WordPress.pbxproj` intact and avoiding possible merge conflicts
* There is no issue with bundle identifiers being changed on build time and having them polluting your git changes (like it happens with the solution in #16095)
* Files added in the Jetpack project will only live there and they will not affect WordPress

### Cons

* If we want to change a screen that already exists in WordPress, we need to add preprocessor macros (`#if JETPACK`) in the original code. Either that or duplicating the file (which might become very complex and costly very quickly)
* If new files get added to the WordPress project, they won't be automatically added to the Jetpack project. We'll have to do it manually or (preferably, and if possible) write a custom script to take care of that for us
